base/parse.jl

```diff
@@ -218,21 +217,36 @@ end
 
 ## string to float functions ##
 
-tryparse(::Type{Float64}, s::String) = ccall(:jl_try_substrtod, Nullable{Float64}, (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
-tryparse(::Type{Float64}, s::SubString{String}) = ccall(:jl_try_substrtod, Nullable{Float64}, (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
-
-tryparse(::Type{Float32}, s::String) = ccall(:jl_try_substrtof, Nullable{Float32}, (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
-tryparse(::Type{Float32}, s::SubString{String}) = ccall(:jl_try_substrtof, Nullable{Float32}, (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
-
+function tryparse(::Type{Float64}, s::String)
+    hasvalue, val = ccall(:jl_try_substrtod, Tuple{Bool, Float64},
+                          (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
+    hasvalue ? Some(val) : nothing
+end
+function tryparse(::Type{Float64}, s::SubString{String})
+    hasvalue, val = ccall(:jl_try_substrtod, Tuple{Bool, Float64},
+                          (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
+    hasvalue ? Some(val) : nothing
+end
+function tryparse(::Type{Float32}, s::String)
+    hasvalue, val = ccall(:jl_try_substrtof, Tuple{Bool, Float32},
+                          (Ptr{UInt8},Csize_t,Csize_t), s, 0, sizeof(s))
+    hasvalue ? Some(val) : nothing
+end
+function tryparse(::Type{Float32}, s::SubString{String})
+    hasvalue, val = ccall(:jl_try_substrtof, Tuple{Bool, Float32},
+                          (Ptr{UInt8},Csize_t,Csize_t), s.string, s.offset, s.endof)
+    hasvalue ? Some(val) : nothing
+end
 tryparse(::Type{T}, s::AbstractString) where {T<:Union{Float32,Float64}} = tryparse(T, String(s))
 
-tryparse(::Type{Float16}, s::AbstractString) = convert(Nullable{Float16}, tryparse(Float32, s))
+tryparse(::Type{Float16}, s::AbstractString) =
+    convert(Union{Some{Float16}, Void}, tryparse(Float32, s))
 
 function parse(::Type{T}, s::AbstractString) where T<:AbstractFloat
     result = tryparse(T, s)
-    if isnull(result)
+    if result === nothing
         throw(ArgumentError("cannot parse $(repr(s)) as $T"))
     end
-    return unsafe_get(result)
+    return get(result)
 end
 
```

Comment:
The `Some` wrappers in this file don't seem necessary

Reply:
I disagree; the `Some` wrappers signal that the result was indeed parsed correctly.

Reply:
The value itself also signals that. There's no additional information being provided by the `Some` wrapper, just extra typing.

Reply:
There's a single case where `Some` would be needed: it's for `parse(Void, s)`. If we don't plan to support this (doesn't make a lot of sense), then I guess we can remove it.

Then one can always make the argument that `Some` forces the caller to handle the possibility of the result being `nothing`. There are advocates of this approach in other languages, though I'm not completely convinced bothering users is useful.

Reply:
> forces the caller to handle the possibility of the result being nothing

Basic test philosophy will also do that. If your code didn't need tests, then it probably also didn't care that much about handling failures either. Also, this is pretty easy to detect as a Lint pass. Also, it's called `tryparse`: it should be pretty obvious to a code-reviewer if you handle it like `parse`. Anyways, I'm aware that there's vast bodies of discussion about this, which I'm neither trying to dismiss or assert here, just stating what's been JuliaLang's typical stance on the matter, since the other approaches do indeed have their merits too.

Reply:
I have to wonder if `tryparse` and `parse` shouldn't be merged with this change.

Reply:
> I have to wonder if tryparse and parse shouldn't be merged with this change.

Maybe, but let's defer this decision to another issue?

Reply:
Actually there are less absurd situations where `Some` might be needed. One might want to have a method like `tryparse(Union{Int, Void}, str)`, which should return `Some(nothing)` when the input is `"nothing"`, and `nothing` when the input is invalid. See https://github.com/JuliaLang/julia/pull/16981.

This situation is less likely now that we no longer conflate the engineer's null and the data scientist's null in a single type as `Nullable` did: we could just state that we don't care about allowing to parse `nothing`. OTOH there may be contexts where it makes sense to parse `"nothing"`, e.g. when parsing JSON files.

Reply:
After considering this question again, I've decided to return `Union{T, Void}` rather than `Union{Some{T}, Void}`. Indeed, the only case when the `Some` wrapper would be useful would be (as in the example above) when passing a `Union` type to `parse`. But that does not sound like an interesting use case. The only reason I can imagine in favor of supporting `parse(Union{T, U}}, s}` would be performance, if the parsing work can be share between the types. Else you can do `x = tryparse(T, s); x === nothing && (x = tryparse(U, s)`. And that performance hypothesis doesn't sound too realistic to me. For `tryparse(Union{T, Void}, s)`, you'd just check `s == "nothing"` or something like that, which is indeed what code does currently.

So I've added a commit to change this.